lib/Runtime/Library/AtomicsObject.cpp

```diff
@@ -0,0 +1,270 @@
+    }
+
+    Var AtomicsObject::EntryExchange(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.exchange");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedExchange(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryIsLockFree(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(1, "Atomics.isLockFree");
+        uint32 size = JavascriptConversion::ToUInt32(args[1], scriptContext);
+        return (size == 1 || size == 2 || size == 4) ? scriptContext->GetLibrary()->GetTrue() : scriptContext->GetLibrary()->GetFalse();
+    }
+
+    Var AtomicsObject::EntryLoad(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(2, "Atomics.load");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedLoad(accessIndex);
+    }
+
+    Var AtomicsObject::EntryOr(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.or");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedOr(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryStore(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.store");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedStore(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntrySub(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.sub");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext);
+        return typedArrayBase->TypedSub(accessIndex, args[3]);
+    }
+
+    Var AtomicsObject::EntryWait(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        ATOMICS_FUNCTION_BEGIN_WRAPPER(3, "Atomics.wait");
+
+        uint32 accessIndex = 0;
+        TypedArrayBase *typedArrayBase = ValidateAndGetTypedArray(args[1], args[2], &accessIndex, scriptContext, true /*onlyInt32*/);
+
+        Var value = args[3];
+        Var timeoutVar = args[4];
+        uint32 timeout = INFINITE;
+
+        if (args.Info.Count > 4 && !JavascriptOperators::IsUndefinedObject(timeoutVar))
+        {
+            double t =JavascriptConversion::ToNumber(timeoutVar, scriptContext);
+            if (!(NumberUtilities::IsNan(t) || JavascriptNumber::IsPosInf(t)))
```

Comment:
Yes, the call to `max` will you give 0. I should have test for that. 

---
In reply to: [77405070](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77405070) [](ancestors = 77405070)